pygmt/datatypes/header.py

```diff
@@ -0,0 +1,228 @@
+"""
+Wrapper for the GMT_GRID_HEADER data structure and related utility functions.
+"""
+
+import ctypes as ctp
+from typing import ClassVar
+
+import numpy as np
+
+# Constants for lengths of grid header variables.
+#
+# Note: Ideally we should be able to get these constants from the GMT shared library
+# using the ``lib["GMT_GRID_UNIT_LEN80"]`` syntax, but it causes cyclic import error.
+# So we have to hardcode the values here.
+GMT_GRID_UNIT_LEN80 = 80
+GMT_GRID_TITLE_LEN80 = 80
+GMT_GRID_COMMAND_LEN320 = 320
+GMT_GRID_REMARK_LEN160 = 160
+
+# GMT uses single-precision for grids by default, but can be built to use
+# double-precision. Currently, only single-precision is supported.
+gmt_grdfloat = ctp.c_float
+
+
+class _GMT_GRID_HEADER(ctp.Structure):  # noqa: N801
+    """
+    GMT grid header structure for metadata about the grid.
+
+    The class is used in the `GMT_GRID`/`GMT_IMAGE`/`GMT_CUBE` data structure. See the
+    GMT source code gmt_resources.h for the original C structure definitions.
+    """
+
+    _fields_: ClassVar = [
+        # Number of columns
+        ("n_columns", ctp.c_uint32),
+        # Number of rows
+        ("n_rows", ctp.c_uint32),
+        # Grid registration, 0 for gridline and 1 for pixel
+        ("registration", ctp.c_uint32),
+        # Minimum/maximum x and y coordinates
+        ("wesn", ctp.c_double * 4),
+        # Minimum z value
+        ("z_min", ctp.c_double),
+        # Maximum z value
+        ("z_max", ctp.c_double),
+        # x and y increments
+        ("inc", ctp.c_double * 2),
+        # Grid values must be multiplied by this factor
+        ("z_scale_factor", ctp.c_double),
+        # After scaling, add this offset
+        ("z_add_offset", ctp.c_double),
+        # Units in x-directions, in the form "long_name [units]"
+        ("x_units", ctp.c_char * GMT_GRID_UNIT_LEN80),
+        # Units in y-direction, in the form "long_name [units]"
+        ("y_units", ctp.c_char * GMT_GRID_UNIT_LEN80),
+        # Grid value units, in the form "long_name [units]"
+        ("z_units", ctp.c_char * GMT_GRID_UNIT_LEN80),
+        # Name of data set
+        ("title", ctp.c_char * GMT_GRID_TITLE_LEN80),
+        # Name of generating command
+        ("command", ctp.c_char * GMT_GRID_COMMAND_LEN320),
+        # Comments for this data set
+        ("remark", ctp.c_char * GMT_GRID_REMARK_LEN160),
```

Comment:
Store these metadata in the `attrs` dict too?

Reply:
Done in dfbc2e5237ff53e6741be2905dcb43a3e3a5877f.